strategies/categorized/src/main/java/org/arquillian/smart/testing/strategies/categorized/CategoriesParser.java

```diff
@@ -0,0 +1,88 @@
+package org.arquillian.smart.testing.strategies.categorized;
+
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Method;
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.arquillian.smart.testing.logger.Log;
+import org.arquillian.smart.testing.logger.Logger;
+
+class CategoriesParser {
+
+    private final Logger logger = Log.getLogger();
+    private final CategorizedConfiguration strategyConfig;
+    private final List<String> specifiedCategories;
+
+    CategoriesParser(CategorizedConfiguration strategyConfig) {
+        this.strategyConfig = strategyConfig;
+        specifiedCategories = Arrays.stream(strategyConfig.getCategories())
+            .map(this::changeIfNonCaseSensitive)
+            .collect(Collectors.toList());
+    }
+
+    boolean hasCorrectCategories(Class<?> clazz) {
+        List<String> presentCategories = getPresentCategories(clazz);
+
+        if (!presentCategories.isEmpty()) {
+            if (specifiedCategories.size() > 0) {
+
+                List<String> intersection = presentCategories.stream()
+                    .filter(category -> isSpecified(category, specifiedCategories))
+                    .collect(Collectors.toList());
+
+                if (strategyConfig.isMatchAll()) {
+                    return intersection.size() == specifiedCategories.size();
+                } else {
+                    return !intersection.isEmpty();
+                }
+            }
+            return true;
+        }
+        return false;
+    }
```

Comment:
This could be as well written as 

```java
final List<String> presentCategories = getPresentCategories(clazz);
        if (presentCategories.isEmpty()) {
            return false;
        }

        final List<String> intersection = presentCategories.stream()
            .filter(category -> isSpecified(category, specifiedCategories))
            .collect(Collectors.toList());

        if (strategyConfig.isMatchAll() || specifiedCategories.isEmpty()) {
            return intersection.size() == specifiedCategories.size();
        } else {
            return !intersection.isEmpty();
        }
```

Which might be a bit simpler to read (at least for me). But this makes me wonder if we shouldn't write some unit test for parser covering boundary cases if we don't cover all of them yet in the feature tests for detector.

Reply:
I see your point with the refactoring - I'll change it with your version. 

As for the test - all boundary cases should be covered by the `CategorizedTestsDetectorTest`. But I do agree that it can be confusing that there is missing the unit test for this class. The reason is that the whole logic was previously inside of the `CategorizedTestsDetector` (when I was writing the test). I moved it to separated class afterward. 
I'll create an additional unit test for it.